polaris/hub/storage.py

```diff
@@ -125,6 +129,149 @@ def _multipart_upload(self, key: str, value: bytes) -> None:
+                    # Setting this header will raise an error if the key already exists.
+                    extra = {"IfNoneMatch": "*"} if if_exists == "skip" else {}
+                    self.s3_client.put_object(
+                        Bucket=self.bucket_name,
+                        Key=self._full_key(key),
+                        Body=data,
+                        ContentType=self.content_type,
+                        ContentMD5=b64encode(md5_hash.digest()).decode(),
+                        Metadata={
+                            "md5sum": md5_hash.hexdigest(),
+                        },
+                        **extra,
+                    )
+                    bytes_copied = buffer_size(data)
+                    copied = 1
+                except ClientError as error:
+                    error_code = error.response["Error"]["Code"]
+                    # Raised when the key already exists and IfNoneMatch is set
+                    if error_code == "PreconditionFailed":
+                        match if_exists:
+                            case "skip":
+                                skipped = 1
+                            case "raise":
+                                raise CopyError(f"key {key!r} exists in destination")
+                    else:
+                        raise
+            return copied, skipped, bytes_copied
+
+        source_store_version = getattr(source, "_store_version", 2)
+        if source_store_version != self._store_version:
+            raise ValueError("zarr stores must share the same protocol version")
+
+        # The executor will allocate min(nbr_CPU + 4, 32) threads
+        with ThreadPoolExecutor() as executor:
+            total_copied = total_skipped = total_bytes_copied = 0
+
+            number_source_keys = len(source)
+
+            # Batch the keys, otherwise we end up with too many files open at the same time
+            batch_iter = iter(source.keys())
+            while batch := tuple(islice(batch_iter, self._batch_size)):
+                # Create a future for each key to copy
+                future_to_key = [
+                    executor.submit(copy_key, source_key, source, if_exists) for source_key in batch
+                ]
+
+                # As each future completes, collect the results
+                for future in as_completed(future_to_key):
+                    result_copied, result_skipped, result_bytes_copied = future.result()
+                    total_copied += result_copied
+                    total_skipped += result_skipped
+                    total_bytes_copied += result_bytes_copied
+
+                log(
+                    f"Copied {total_copied} ({total_bytes_copied / (1024**2):.2f} MiB), skipped {total_skipped}, of {number_source_keys} keys. {(total_copied + total_skipped) / number_source_keys * 100:.2f}% completed."
+                )
+
+            return total_copied, total_skipped, total_bytes_copied
```

Comment:
This is the kind of code that makes you think, "Maybe I could do this for a living and actually enjoy it." 😅